### PR TITLE
Sentence classification - Doc - Example

### PR DESCRIPTION
**Pull request by raphaelsty**
Hi team,

Here is a small pull-request to add an example in the doc. I created a notebook dedicated to sentence classification using the `SMSSpam` dataset. This tutorial is related to the discussion #482.

Goals of this tutorial:

- Handle text using river
- How to use `naive_bayes` using TFIDF.
- How to use regression logistic for classification using TFIDF
- Because distributions of targets of SMSSPAM dataset are not uniform, some links with the `imblearn` tutorial and module.
- But mainly **how to integrate pre-trained word embeddings (spicy) into river pipelines.**

Sorry, I did not ask if it was useful to do it but I think it's great to have more examples in the documentation.

Thank you in advance for the review.

Raphaël

### CODE REVIEW

File: docs/examples/sentence_classification.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
+# Sentence classification
+
+In this tutorial we will try to predict whether a sms is spam or not. To train our model, we will use the `SMSSpam` dataset. This dataset is unbalanced, there is only 13.4% spam. Let's look at the data:
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:3]:
```suggestion
In this tutorial we will try to predict whether a sms is spam or not. To train our model, we will use the `SMSSpam` dataset. This dataset is unbalanced, there is only 13.4% spam. Let's look at the data:
```
```suggestion
In this tutorial we will try to predict whether an SMS is a spam or not. To train our model, we will use the `SMSSpam` dataset. This dataset is unbalanced, there is only 13.4% spam. Let's look at the data:
```

File: docs/examples/sentence_classification.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+
+
+```python
+from pprint import pprint
+
+X_y = datasets.SMSSpam()
+
+for x, y in X_y:
+
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:39]: I know you well enough now to know that you like adding a lot of spaces between lines of code. It's a good reflex. However I feel that in this tutorial you went a bit far :). In this particular case I would say that there's no need for a space. But it's your notebook so your decision to make! :)

File: docs/examples/sentence_classification.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    print(f'Spam: {y}')
+    break
+```
+
+    {'body': 'Go until jurong point, crazy.. Available only in bugis n great world '
+             'la e buffet... Cine there got amore wat...\n'}
+    Spam: False
+
+
+Let's start by building a simple model like a naive bayes classifier. We will first transform the sentences with a tf-idf to feed our model. Then, we will measure the accuracy of our model with the AUC metric. This metric is adapted when the classes of our dataset are not uniformly distributed. In addition, the Naive Bayes models can perform very well on unbalanced datasets and supports multiple classes for multiple classification problems.
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:50]:
```suggestion
Let's start by building a simple model like a Naive Bayes classifier. We will first preprocess the sentences with a TF-IDF transform that our model can consume. Then, we will measure the accuracy of our model with the AUC metric. This is the right metric to use when the classes are not balanced. In addition, the Naive Bayes models can perform very well on unbalanced datasets and can be used for both binary and multi-class classification problems.
```

File: docs/examples/sentence_classification.md [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+               False    True
+       False    4809      17
+        True     103     644
+
+
+
+The results are quite good with this first model.
+
+Since we are working with an unbalanced dataset, we can use the `imblearn` module to artificially increase the number of spam in our dataset. For more information about the `imblearn` module, you can find a dedicated tutorial [here](https://riverml.xyz/latest/examples/imbalanced-learning/).
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:110]: You shouldn't use a hard link. You should reference relative to where you are. So in this case it would be something like `[here](./imbalanced-learning)`. I'm not exactly sure so I'll let you try locally.

File: docs/examples/sentence_classification.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+from river import naive_bayes
+from river import metrics
+
+def extract_body(x):
+    """Extract the body of the sms."""
+    return x['body']
+
+X_y = datasets.SMSSpam()
+
+model = extract_body | feature_extraction.TFIDF() | naive_bayes.BernoulliNB(alpha = 0)
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:64]:
```suggestion
model = (
    extract_body |
    feature_extraction.TFIDF() |
    naive_bayes.BernoulliNB(alpha=0)
)
```

File: docs/examples/sentence_classification.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+```python
+from river import imblearn
+
+X_y = datasets.SMSSpam()
+
+model = extract_body | feature_extraction.TFIDF() | imblearn.RandomUnderSampler(
+    classifier=naive_bayes.BernoulliNB(alpha = 0),
+    desired_dist={0: 0.5, 1: 0.5},
+    seed=42
+)
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:122]:
```suggestion
model = (
    extract_body |
    feature_extraction.TFIDF() |
    imblearn.RandomUnderSampler(
        classifier=naive_bayes.BernoulliNB(alpha=0),
        desired_dist={0: 0.5, 1: 0.5},
        seed=42
    )
)
```

File: docs/examples/sentence_classification.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+
+
+    
+![svg](sentence_classification_files/sentence_classification_14_0.svg)
+    
+
+
+
+Now let's try to use logistic regression to classify messages. I will use different tips to make my model perform better. As in the previous example, I artificially increase the number of spam. The logistics regression will be fed from a tf-idf.
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:182]: I think it's better if you use the "we" pronoun instead of "I". It seems like a convention in technical documention. What do you think?

Also there's a typo at "logistics" and you need to capitalize "tf-idf".

File: docs/examples/sentence_classification.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+from river import optim
+from river import preprocessing
+
+X_y = datasets.SMSSpam()
+
+model = extract_body | feature_extraction.TFIDF() | preprocessing.StandardScaler() | imblearn.RandomUnderSampler(
+    classifier=linear_model.LogisticRegression(optimizer=optim.SGD(0.05), loss=optim.losses.Log()),
+    desired_dist={0: 0.5, 1: 0.5},
+    seed=42
+)
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:196]:
```suggestion
model = (
    extract_body |
    feature_extraction.TFIDF() |
    preprocessing.StandardScaler() |
    imblearn.RandomUnderSampler(
        classifier=linear_model.LogisticRegression(
            optimizer=optim.SGD(0.05),
            loss=optim.losses.Log()
        ),
        desired_dist={0: 0.5, 1: 0.5},
        seed=42
    )
)
```

File: docs/examples/sentence_classification.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+Now let's try to use logistic regression to classify messages. I will use different tips to make my model perform better. As in the previous example, I artificially increase the number of spam. The logistics regression will be fed from a tf-idf.
+
+
+```python
+from river import linear_model
+from river import optim
+from river import preprocessing
+
+X_y = datasets.SMSSpam()
+
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:191]: I think using a `Normalizer` instead of a `StandardScaler` is more appropriate. At least this is what scikit-learn recommends. Indeed the goal is to normalize the length of documents, and not necessarily the distribution of a feature's occurrences. In other words it should be better to normalize row-wise rather than column-wise.

File: docs/examples/sentence_classification.md [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+    
+![svg](sentence_classification_files/sentence_classification_19_0.svg)
+    
+
+
+
+The results of the logistic regression are less good. It is more difficult for the model to determine if a text message is spam.
+
+Let's try to use word embeddings to improve our logistic regression. Word embeddings allow you to represent a word as a vector. Embeddings are developed to build semantic rich vectors, i.e. the vector which represents the word **python** is close to the vector which represents the word **programming**. We will use Scipy to convert our sentence to vectors. Scipy convert a sentence to a vector by calculating the average of the embeddings of the words of the sentence.
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:254]:
```suggestion
Let's try to use word embeddings to improve our logistic regression. Word embeddings allow you to represent a word as a vector. Embeddings are developed to build semantically rich vectors. For instance, the vector which represents the word **python** should be close to the vector which represents the word **programming**. We will use [spaCy](https://spacy.io/) to convert our sentence to vectors. spaCy converts a sentence to a vector by calculating the average of the embeddings of the words in the sentence.
```

File: docs/examples/sentence_classification.md [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+You can download pre-trained embeddings in many languages. We will use English pre-trained embeddings as our sms are in English.
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:256]:
```suggestion
You can download pre-trained embeddings in many languages. We will use English pre-trained embeddings as our SMS are in English.
```

File: docs/examples/sentence_classification.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+
+The command below allows you to download the embeddings of spacy. More informations about spacy and it's installation [here](https://spacy.io/usage).
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:258]:
```suggestion
The command below allows you to download the pre-trained embeddings that spaCy makes available. More informations about spaCy and its installation may be found here [here](https://spacy.io/usage).
```

File: docs/examples/sentence_classification.md [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+python -m spacy download en_core_web_sm
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:259]:
```suggestion
```sh
python -m spacy download en_core_web_sm
```
```

File: docs/examples/sentence_classification.md [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+from river.base import Transformer
+
+class Embeddings(Transformer):
+    """My custom transformer, words embeddings using spacy."""
+    
+    def __init__(self):
+        self.embeddings = spacy.load('en_core_web_sm')
+        
+    def transform_one(self, x, y=None):
+        return {i: xi for i, xi in enumerate(self.embeddings(x).vector)}
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:275]: I would give more meaning to `i`, for instance call it `dimension` or something like that.

File: docs/examples/sentence_classification.md [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+
+
+```python
+X_y = datasets.SMSSpam()
+
+model = extract_body | Embeddings() | preprocessing.StandardScaler() | imblearn.RandomOverSampler(
+    classifier=linear_model.LogisticRegression(loss=optim.losses.Log()),
+    desired_dist={0: 0.5, 1: 0.5},
+    seed=42
+)
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:288]:
```suggestion
model = (
    extract_body |
    Embeddings() |
    preprocessing.StandardScaler() |
    imblearn.RandomOverSampler(
        classifier=linear_model.LogisticRegression(
            loss=optim.losses.Log()
        ),
        desired_dist={0: 0.5, 1: 0.5},
        seed=42
    )
)
```

File: docs/examples/sentence_classification.md [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+
+
+
+    
+![svg](sentence_classification_files/sentence_classification_31_0.svg)
+    
+
+
+
+The results of the logistic regression with the spacy embeddings are much better than the tf-idf. We could surely improve the results by cleaning up the text. However, on this problem, the logistic regression is not better than the naive bayes model. No free lunch today.
```
Comment (MaxHalford) [inline docs/examples/sentence_classification.md:344]:
```suggestion
The results of the logistic regression with the spacy embeddings are much better than the tf-idf. We could surely improve the results by cleaning up the text. However, on this problem, the logistic regression is not better than the naive bayes model. No free lunch today.
```
```suggestion
The results of the logistic regression with the spaCy embeddings are much better than when using TF-IDF valies. We could surely improve the results by cleaning up the text. However, on this problem, the logistic regression is not better than the Naive Bayes model. No free lunch today.
```

**Review by MaxHalford — COMMENTED** [review]
I'm done reviewing :). Really good job on this, it's super cool. I think it's a great example.

**Comment by raphaelsty** [comment]
@MaxHalford Again, thank you for the review, I followed all your comments and updated the code and comments. The link to the `imblearn` tutorial works fine locally. By replacing standard scaler with normalize, we improved the results of the regression logistic + TF-IDF. I made some updates on the description.
